# NanoporeFiltrationAMR
Runs KMA to align sequences in the (Nanopore) input fastq.gz file to the ResFinder database of antimicrobial resistance genes.
To remove gene-hits due to contamination and sequencing errors, two filtrations based on the depth of gene-hits are applied.
Results after the filtration are outputted in plaintext format.

## Requirements

- [KMA](https://bitbucket.org/genomicepidemiology/kma/src/master/) 

- Indexed ResFinder database. Created by running KMA index on [ResFinder 4.0](https://bitbucket.org/genomicepidemiology/resfinder_db/src/master/)
- phenotypes.txt file from the ResFinder database.
- Nanopore reads in fastq.gz format

## Usage

To run full pipeline:

```
python3 NanoporeFiltrationAMR.py [-i <Nanopore fastq.gz file>] [-o <Output folder>] [-gs <Genome size>] [-pf <Phenotypes.txt file from ResFinder database>] [-refdb <Reference database generated by KMA index>]
```

## Input arguments
Required arguments
```
-i <Path to Nanopore fastq.gz file>
-o <Path to output folder>
-gs <Genome size - Full number or affixed by K,M or G>
-pf <Path to the phenotypes.txt file from the ResFinder database. Used for genotypic to phenotypic translation>
-refdb <Path to reference database, generated by running KMA index on the ResFinder database>
```

Optional arguments
```
-cov <Set the minimum coverage threshold for including genes. Value between 0 and 100. Default 90.>
-id <Set the minimum identity threshold for including genes. Value between 0 and 100. Default 90.>
-md <Set the minimum depth for gene-groups to be included. Value between 0.0 and 1.0. Default 0.2>
-gd <Set the minimum depth for gene-variants, based on depth relative to highest depth gene from the same gene-group. Value between 0.0 and 1.0. Default 0.45>
```
